### PR TITLE
feat: fork ament_cmake_auto package for ament/ament_cmake#385

### PR DESCRIPTION
**Pull request by kenji-miyake**
## Description

I caught the following error in https://github.com/autowarefoundation/autoware/pull/330.

```
Starting >>> interpolation
[Processing: interpolation]
--- stderr: interpolation
In file included from /usr/include/eigen3/Eigen/Core:214,
                 from /autoware/install/tier4_autoware_utils/include/tier4_autoware_utils/geometry/boost_geometry.hpp:23,
                 from /autoware/install/tier4_autoware_utils/include/tier4_autoware_utils/geometry/geometry.hpp:23,
                 from /autoware/src/universe/autoware.universe/common/interpolation/include/interpolation/spline_interpolation.hpp:19,
                 from /autoware/src/universe/autoware.universe/common/interpolation/include/interpolation/spline_interpolation_points_2d.hpp:18,
                 from /autoware/src/universe/autoware.universe/common/interpolation/src/spline_interpolation_points_2d.cpp:15:
/usr/include/eigen3/Eigen/src/Core/arch/NEON/PacketMath.h: In function ‘Packet Eigen::internal::pload(const typename Eigen::internal::unpacket_traits<T>::type*) [with Packet = Eigen::internal::eigen_packet_wrapper<int, 2>; typename Eigen::internal::unpacket_traits<T>::type = signed char]’:
/usr/include/eigen3/Eigen/src/Core/arch/NEON/PacketMath.h:1671:9: error: ‘void* memcpy(void*, const void*, size_t)’ copying an object of non-trivial type ‘Eigen::internal::Packet4c’ {aka ‘struct Eigen::internal::eigen_packet_wrapper<int, 2>’} from an array of ‘const int8_t’ {aka ‘const signed char’} [-Werror=class-memaccess]
 1671 |   memcpy(&res, from, sizeof(Packet4c));
      |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

This is the same reason as https://github.com/ANYbotics/grid_map/pull/349 and only happens with `arm64`.

Since we use `ament_cmake_auto`, I believe we can't apply the same solution for this until https://github.com/ament/ament_cmake/pull/385 is merged and released.

Therefore, I'd like to add common settings in `autoware_cmake`.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kenji-miyake** [comment]
@esteve @xmfcx Could you please review this? :bow: :

File: autoware_cmake/cmake/autoware_package.cmake [hunk 1]
```diff
@@ -44,6 +44,12 @@ macro(autoware_package)
     add_compile_definitions(ROS_DISTRO_HUMBLE)
   endif()
 
+  # Set common system includes
```
Comment (esteve) [inline autoware_cmake/cmake/autoware_package.cmake:47]: I think this is not the right place for this, this means that every package will call `find_package` even if they don't need Eigen. I'd update the packages that actually need Eigen to call `include_directories(SYSTEM...) `

**Review by esteve — CHANGES_REQUESTED** [review]

File: autoware_cmake/cmake/autoware_package.cmake (hunk 1 above)
Comment (kenji-miyake) [reply autoware_cmake/cmake/autoware_package.cmake:47]: I can remove this if https://github.com/ament/ament_cmake/pull/385 is released.

Comment (esteve) [reply autoware_cmake/cmake/autoware_package.cmake:47]: I think we can manually call `include_directories(SYSTEM ...)` in those packages that depend on Eigen, even if we are using `ament_cmake_auto`. I'm going to test this.

Comment (kenji-miyake) [reply autoware_cmake/cmake/autoware_package.cmake:47]: The [interpolation package](https://github.com/autowarefoundation/autoware.universe/tree/main/common/interpolation)
 has an error, but it doesn't depend on Eigen.

Comment (esteve) [reply autoware_cmake/cmake/autoware_package.cmake:47]: Even more reason for not calling `find_package(Eigen ...)` if the interpolation package does not depend on Eigen, adding that to `autoware_package` would be too much, every package would include the Eigen headers no matter what.

**Comment by xmfcx** [comment]
@kenji-miyake wrote:
> I can remove this if https://github.com/ament/ament_cmake/pull/385 is released.

Would it work if we put our fork of ament_cmake with that PR merged into autoware? Would it make sense?

**Comment by kenji-miyake** [comment]
@xmfcx I think it's possible and reasonable. But I'm not sure what to do if the PR is rejected.

**Comment by xmfcx** [comment]
> @xmfcx I think it's possible and reasonable. But I'm not sure what to do if the PR is rejected.

It has an approval, I am optimistic about its merging to rolling. But I am guessing we won't see it in humble.

**Comment by kenji-miyake** [comment]
Okay, then I'll add the package instead! Please review that later.

**Comment by kenji-miyake** [comment]
I cherry-picked three commits into Humble.
image

And I've executed the following commands:

```bash
git filter-repo --path ament_cmake_auto --force
git remote add origin ssh://git@github.com/autowarefoundation/autoware_common.git
git fetch
git rebase -i origin/set-common-system-headers # squashed all commits into one
git revert c77fc31557f6b5bbb13e4390683a755a426378c4
```

**Comment by kenji-miyake** [comment]
@xmfcx @esteve I've tested this PR with the following commands, and it worked.

```bash
docker run --rm -it ghcr.io/autowarefoundation/autoware-universe:humble-latest-arm64
cd src/core/autoware_common
git fetch git checkout set-common-system-headers
cd /autoware
colcon build --packages-up-to interpolation
```

**Comment by kenji-miyake** [comment]
Either one is okay for me.

**Comment by kenji-miyake** [comment]
If we copy all of `ament_cmake`, the number of packages will increase.
But even if we split the repository (= normally fork), we can remove other unnecessary packages from the fork like this PR.

**Pull request closed without merging**